app/controllers/ems_infra_controller.rb

```diff
@@ -33,52 +33,79 @@ def scaling
       return
     end
 
-    @count_parameters = @stack.parameters.select { |x| x.name.include?('::count') }
+    arel_table = OrchestrationStackParameter.arel_table
+    @orig_parameters = @stack.parameters.where(arel_table[:name].matches("%::count"))
 
     return unless params[:scale]
 
-    scale_parameters = params.select { |k, _v| k.include?('::count') }
-    assigned_hosts = scale_parameters.values.sum(&:to_i)
+    form_parameters = params.select { |k, _v| k.include?('::count') }
+
+    validate_selected_hosts_not_exceed_available_hosts(form_parameters)
+
+    changed_parameters = select_changed_parameters(form_parameters)
+    validate_a_change_was_made(changed_parameters)
+
+    return unless changed_parameters.length > 0
+
+    begin
+      # Check if stack is ready to be updated
+      update_ready = @stack.update_ready?
+    rescue => ex
+      log_and_flash_message(_("Unable to initiate scaling, obtaining of status failed: #{ex}"))
+      return
+    end
+
+    unless update_ready
+      add_flash(_("Provider is not ready to be scaled, another operation is in progress."), :error)
+    end
+
+    begin
+      @stack.raw_update_stack(:parameters => changed_parameters)
+      redirect_to :action => 'show', :id => params[:id], :flash_msg => select_return_message(form_parameters)
+    rescue => ex
+      log_and_flash_message(_("Unable to initiate scaling: %s") % ex)
+    end
+  end
+
+  private ############################
+
+  def value_changed?(form_parameters, name, value)
+    !form_parameters[name].nil? && form_parameters[name] != value
+  end
+
+  def select_changed_parameters(form_parameters)
+    changed_parameters = []
+    @orig_parameters.each do |orig|
+      next unless value_changed?(form_parameters, orig.name, orig.value)
+      changed_parameters << {"name" => orig.name, "value" => form_parameters[orig.name]}
+    end
+    changed_parameters
+  end
+
+  def select_return_message(form_parameters)
```

Comment:
Or suggestion works. I find it more confusing, but that is just me not familiar with that style. :)